### PR TITLE
gateway wildcard hostname e2e test

### PR DESCRIPTION
**Pull request by stevesloka**
Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

File: test/e2e/gateway/009_tls_wildcard_host_test.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+func testTLSWildcardHost(fx *e2e.Framework) {
+	t := fx.T()
+	namespace := "009-gateway-tls-wildcard-host"
+	hostSuffix := "wildcardhost.gateway.projectcontour.io"
+
+	fx.CreateNamespace(namespace)
+	defer fx.DeleteNamespace(namespace)
+
+	fx.Fixtures.Echo.Deploy(namespace, "echo")
+	fx.Certs.CreateSelfSignedCert(namespace, "echo-one-cert", "echo-one-cert", "*."+hostSuffix)
```
Comment (skriss) [inline test/e2e/gateway/009_tls_wildcard_host_test.go:37]: Does this cert need to be configured on the Gateway? If so, see https://github.com/projectcontour/contour/blob/5e45ddc6a35b95810d5d3e86f3dd91d83f10ccf0/test/e2e/gateway/gateway_test.go#L144-L150 and https://github.com/projectcontour/contour/blob/5e45ddc6a35b95810d5d3e86f3dd91d83f10ccf0/test/e2e/gateway/gateway_test.go#L166 for how that's done. You'd need a new [Describe](https://github.com/projectcontour/contour/blob/5e45ddc6a35b95810d5d3e86f3dd91d83f10ccf0/test/e2e/gateway/gateway_test.go#L113) block for the Gateway with the wildcard cert

File: test/e2e/gateway/009_tls_wildcard_host_test.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+		t.Logf("Making request with hostname=%s, sni=%s", tc.hostname, tc.sni)
+
+		res, ok := fx.HTTP.SecureRequestUntil(&e2e.HTTPSRequestOpts{
+			Host: tc.hostname,
+			TLSConfigOpts: []func(*tls.Config){
+				e2e.OptSetSNI(tc.sni),
+			},
+			Condition: e2e.HasStatusCode(tc.wantStatus),
+		})
+		require.Truef(t, ok, "expected %d response code, got %d", tc.wantStatus, res.StatusCode)
```
Comment (sunjayBhatia) [inline test/e2e/gateway/009_tls_wildcard_host_test.go:119]: panic might be resolved by the cert/tls being set up on the gateway, not sure if we should do a "require" `res` to be non-nil so we can see the error w/o panics

**Review by sunjayBhatia — COMMENTED** [review]
LGTM generally once the gateway/panic is fixed 👍🏽 

curious to see if the `a.random3` test passes, my guess is we will need the same header match hack as for Ingress

**Comment by stevesloka** [comment]
The a.random3 test indeed fails @sunjayBhatia. =)

I was using this PR to run the tests since they don't work on my m1 mac. Need to go look at your wildcard hostname hack to fix this up.

**Review by sunjayBhatia — APPROVED** [review]
LGTM

I wonder if a future refactor could be done to abstract the header match needed for the wildcard host more, i didnt do it at first for Ingress so we could see if doing it again for another resource made things unwieldy

- maybe at the very least a helper function that generates the regex match rule so we don't have to repeat much code?
- make `dag.Route` wildcard aware or only add the header match rule on conversion to Envoy type?
  - would have to take care to consider sorting of routes if we change things

**Pull request merged**